### PR TITLE
Support migration.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/cli/cmds/fleet.toit [hunk 1]
```diff
@@ -429,6 +429,74 @@ create-fleet-commands config/Config cache/Cache ui/Ui -> List:
...
 
+  migration-cmd := cli.Command "migration"
+      --aliases=["migrate"]
+      --help="""
+        Migrate a fleet to a new broker.
+
+        - Add a new broker to your configuration. Use 'config broker add ...'.
+        - Start the migration with 'migration start --broker=<broker>'.
+        - Upload a new pod, and roll it out to the fleet with 'fleet roll-out'.
+        - Check the status of the migration with the 'status' command.
```
Comment (kasperl) [inline src/cli/cmds/fleet.toit:440]:
```suggestion
        - Check the status of the migration with the 'migration status' command.
```

File: src/cli/cmds/fleet.toit [hunk 2]
```diff
@@ -429,6 +429,74 @@ create-fleet-commands config/Config cache/Cache ui/Ui -> List:
...
+          """
+      --options=[
+        cli.Option "broker"
+            --help="The broker to remove."
+            --multi,
+      ]
+      --examples=[
+        cli.Example """
+            Finish the migration.
+            All old brokers are removed, and devices can only b
```
Comment (kasperl) [inline src/cli/cmds/fleet.toit:489]: Incomplete sentence.

File: src/cli/cmds/fleet.toit [hunk 3]
```diff
@@ -751,3 +819,18 @@ create-reference parsed/cli.Parsed config/Config cache/Cache ui/Ui:
...
     fleet.write-reference --path=output
     ui.info "Created reference file $output."
+
+migration-start parsed/cli.Parsed config/Config cache/Cache ui/Ui:
+  broker-name := parsed["broker"]
+
+  with-devices-fleet parsed config cache ui: | fleet/FleetWithDevices |
+    new-broker := get-server-from-config config ui --name=broker-name
+    fleet.migration-start --broker-config=new-broker
+    ui.info "Started migration to broker $broker-name."
```
Comment (kasperl) [inline src/cli/cmds/fleet.toit:829]: Maybe add information on how to check the status right here?

File: src/cli/cmds/fleet.toit [hunk 4]
```diff
@@ -751,3 +819,18 @@ create-reference parsed/cli.Parsed config/Config cache/Cache ui/Ui:
...
+
+migration-finish parsed/cli.Parsed config/Config cache/Cache ui/Ui:
+  brokers := parsed["broker"]
+
+  with-devices-fleet parsed config cache ui: | fleet/FleetWithDevices |
+    fleet.migration-finish brokers
+    ui.info "Finished migration."
```
Comment (kasperl) [inline src/cli/cmds/fleet.toit:836]: Add broker information to the output?

File: src/cli/fleet.toit [hunk 5]
```diff
@@ -122,28 +126,82 @@ class FleetFile:
           ui.abort "Fleet file $path has invalid format for 'pod' in group '$group-name'."
         PodReference.parse entry["pod"] --ui=ui
 
-    broker-config := default-broker-config
-
     broker-name := fleet-content.get "broker"
+    migrated-from-entry := fleet-content.get "migrated-from"
```
Comment (kasperl) [inline src/cli/fleet.toit:130]: Maybe this should be more like 'migrating-from' since it isn't finished yet?

File: src/cli/fleet.toit [hunk 6]
```diff
@@ -560,23 +623,42 @@ class FleetWithDevices extends Fleet:
...
       pods-per-group.get group-name --init=: download (pod-reference-for-group group-name)
 
-    broker.roll-out --devices=existing-devices.values --pods=pods --diff-bases=diff-bases
+    broker.roll-out --devices=detailed-devices.values --pods=pods --diff-bases=diff-bases
 
     ui_.info "Successfully updated $(fleet-devices.size) device$(fleet-devices.size == 1 ? "" : "s")."
 
+    // We need to notify the migrated-from brokers.
+    fleet-file_.migrated-from.do: | server-name |
+      ui_.info "Rolling out on migrated-from broker $server-name."
```
Comment (kasperl) [inline src/cli/fleet.toit:645]: migrated-from sounds so low-level here. Maybe reword it?

Rolling out to $old broker (migration in progress).

File: src/cli/fleet.toit [hunk 7]
```diff
@@ -560,23 +623,42 @@ class FleetWithDevices extends Fleet:
...
+          --ui=ui_
+          --fleet-id=id
+          --organization-id=organization-id
+          --tmp-directory=artemis.tmp-directory
+      // We could filter out devices that were already known in the new broker, but
+      // it's easier and more robust to update all devices.
+      // This also makes it possible to move forward and backward between two brokers.
+      detailed-devices = old-broker.get-devices --device-ids=device-ids
+      old-broker.roll-out --devices=detailed-devices.values --pods=pods --diff-bases=diff-bases
+      ui_.info "Successfully rolled out on migrated-from broker $server-name."
```
Comment (kasperl) [inline src/cli/fleet.toit:660]: Successfully rolled out to $old broker (migration in progress).

File: src/cli/cmds/fleet.toit [hunk 8]
```diff
@@ -429,6 +429,74 @@ create-fleet-commands config/Config cache/Cache ui/Ui -> List:
...
+  migration-cmd := cli.Command "migration"
+      --aliases=["migrate"]
+      --help="""
+        Migrate a fleet to a new broker.
+
+        - Add a new broker to your configuration. Use 'config broker add ...'.
+        - Start the migration with 'migration start --broker=<broker>'.
+        - Upload a new pod, and roll it out to the fleet with 'fleet roll-out'.
+        - Check the status of the migration with the 'status' command.
+        - Finish the migration with 'migration finish', once all devices have migrated.
```
Comment (kasperl) [inline src/cli/cmds/fleet.toit:441]: Maybe this should be stop? Finish seems to imply that we're done with it better than stop, but on the other hand, we don't really know if we're done. If finish checked the status and complained if it wasn't really finished (unless --force), then maybe finish would be better.

All in all, I think start/stop is likely the simpler pair and we still check that we don't stop an incomplete migration without a '--force' flag.

**Review by kasperl — APPROVED** [review]
Awesome stuff!

File: src/cli/cmds/fleet.toit (hunk 1 above)
Comment (floitsch) [reply src/cli/cmds/fleet.toit]: There is no specific `migration status` command. The normal `status` command is enough.
Changed it to "fleet's 'status' command".

File: src/cli/cmds/fleet.toit (hunk 8 above)
Comment (floitsch) [reply src/cli/cmds/fleet.toit]: done.

File: src/cli/cmds/fleet.toit (hunk 2 above)
Comment (floitsch) [reply src/cli/cmds/fleet.toit]: done.

File: src/cli/cmds/fleet.toit (hunk 3 above)
Comment (floitsch) [reply src/cli/cmds/fleet.toit]: done.

File: src/cli/cmds/fleet.toit (hunk 4 above)
Comment (floitsch) [reply src/cli/cmds/fleet.toit]: done.

File: src/cli/fleet.toit (hunk 5 above)
Comment (floitsch) [reply src/cli/fleet.toit]: done.

File: src/cli/fleet.toit (hunk 6 above)
Comment (floitsch) [reply src/cli/fleet.toit]: I like that.
done.

File: src/cli/fleet.toit (hunk 7 above)
Comment (floitsch) [reply src/cli/fleet.toit]: done.

**Review by kasperl — APPROVED** [review]
Looks good!
